spacy/language.py

```diff
@@ -2033,8 +2034,21 @@ def replace_listeners(
             # Go over the listener layers and replace them
             for listener in pipe_listeners:
                 new_model = tok2vec_model.copy()
-                if "replace_listener" in tok2vec_model.attrs:
-                    new_model = tok2vec_model.attrs["replace_listener"](new_model)
+                replace_listener_func = tok2vec_model.attrs.get("replace_listener")
+                if replace_listener_func is not None:
+                    # Pass the extra args to the callback without breaking compatibility with
+                    # old library versions that only expect a single parameter.
+                    num_params = len(
+                        inspect.signature(replace_listener_func).parameters
+                    )
+                    assert num_params in (
+                        1,
+                        3,
+                    ), f"'replace_listener' callback expects {num_params} parameters, but we only support one or three"
```

Comment:
To come back to Daniël's suggestion, I'd make this an actual exception and raise an error instead of the assert-with-message. If it's user-facing, a proper exception feels more appropriate. We can just put this at the end of the `if-else` structure as the fall-back case in which we raise an exception.

Reply:
Changed to an exception.